pip install pgraph - AssertionError: Multiple .dist-info directories

- Pip version: 8.1.0
- Python version: 2.7.11+
- Operating System: Linux lab 4.3.0-kali1-686-pae
### Description:

Please try to install **pgraph** with command: `sudo pip install pgraph --upgrade`, 
### What I've run:

```
root@lab:~/# uname -a 
Linux lab 4.3.0-kali1-686-pae #1 SMP Debian 4.3.5-1kali1 (2016-02-11) i686 GNU/Linux

root@lab:~/# pip install pgraph --upgrade
Collecting pgraph
Requirement already up-to-date: pyramid-chameleon in /usr/local/lib/python2.7/dist-packages (from pgraph)
Requirement already up-to-date: pyramid in /usr/local/lib/python2.7/dist-packages (from pgraph)
Requirement already up-to-date: webtest in /usr/local/lib/python2.7/dist-packages (from pgraph)
Requirement already up-to-date: waitress in /usr/local/lib/python2.7/dist-packages (from pgraph)
Requirement already up-to-date: mock in /usr/local/lib/python2.7/dist-packages (from pgraph)
Collecting py-deps[memcache]>=0.5.5 (from pgraph)
Requirement already up-to-date: celery in /usr/local/lib/python2.7/dist-packages (from pgraph)
Requirement already up-to-date: pyramid-debugtoolbar in /usr/local/lib/python2.7/dist-packages (from pgraph)
Requirement already up-to-date: funcsigs in /usr/local/lib/python2.7/dist-packages (from pgraph)
Requirement already up-to-date: Chameleon in /usr/local/lib/python2.7/dist-packages (from pyramid-chameleon->pgraph)
Requirement already up-to-date: setuptools in /usr/local/lib/python2.7/dist-packages (from pyramid->pgraph)
Requirement already up-to-date: repoze.lru>=0.4 in /usr/local/lib/python2.7/dist-packages (from pyramid->pgraph)
Requirement already up-to-date: zope.interface>=3.8.0 in /usr/lib/python2.7/dist-packages (from pyramid->pgraph)
Requirement already up-to-date: PasteDeploy>=1.5.0 in /usr/local/lib/python2.7/dist-packages (from pyramid->pgraph)
Requirement already up-to-date: WebOb>=1.3.1 in /usr/local/lib/python2.7/dist-packages (from pyramid->pgraph)
Requirement already up-to-date: translationstring>=0.4 in /usr/local/lib/python2.7/dist-packages (from pyramid->pgraph)
Requirement already up-to-date: zope.deprecation>=3.5.0 in /usr/local/lib/python2.7/dist-packages (from pyramid->pgraph)
Requirement already up-to-date: venusian>=1.0a3 in /usr/local/lib/python2.7/dist-packages (from pyramid->pgraph)
Requirement already up-to-date: six in /usr/lib/python2.7/dist-packages (from webtest->pgraph)
Requirement already up-to-date: beautifulsoup4 in /usr/lib/python2.7/dist-packages (from webtest->pgraph)
Requirement already up-to-date: pbr>=0.11 in /usr/local/lib/python2.7/dist-packages (from mock->pgraph)
Requirement already up-to-date: networkx in /usr/local/lib/python2.7/dist-packages (from py-deps[memcache]>=0.5.5->pgraph)
Requirement already up-to-date: wheel in /usr/local/lib/python2.7/dist-packages (from py-deps[memcache]>=0.5.5->pgraph)
Requirement already up-to-date: pip>=6.0.1 in /usr/local/lib/python2.7/dist-packages (from py-deps[memcache]>=0.5.5->pgraph)
Requirement already up-to-date: pylibmc in /usr/local/lib/python2.7/dist-packages (from py-deps[memcache]>=0.5.5->pgraph)
Requirement already up-to-date: kombu<3.1,>=3.0.34 in /usr/local/lib/python2.7/dist-packages (from celery->pgraph)
Requirement already up-to-date: pytz>dev in /usr/lib/python2.7/dist-packages (from celery->pgraph)
Requirement already up-to-date: billiard<3.4,>=3.3.0.23 in /usr/local/lib/python2.7/dist-packages (from celery->pgraph)
Requirement already up-to-date: pyramid-mako>=0.3.1 in /usr/local/lib/python2.7/dist-packages (from pyramid-debugtoolbar->pgraph)
Collecting Pygments (from pyramid-debugtoolbar->pgraph)
  Using cached Pygments-2.1.3-py2.py3-none-any.whl
Collecting decorator>=3.4.0 (from networkx->py-deps[memcache]>=0.5.5->pgraph)
  Using cached decorator-4.0.9-py2.py3-none-any.whl
Requirement already up-to-date: anyjson>=0.3.3 in /usr/local/lib/python2.7/dist-packages (from kombu<3.1,>=3.0.34->celery->pgraph)
Requirement already up-to-date: amqp<2.0,>=1.4.9 in /usr/local/lib/python2.7/dist-packages (from kombu<3.1,>=3.0.34->celery->pgraph)
Collecting Mako>=0.8 (from pyramid-mako>=0.3.1->pyramid-debugtoolbar->pgraph)
Requirement already up-to-date: MarkupSafe>=0.9.2 in /usr/lib/python2.7/dist-packages (from Mako>=0.8->pyramid-mako>=0.3.1->pyramid-debugtoolbar->pgraph)
Installing collected packages: py-deps, pgraph, Pygments, decorator, Mako
Exception:
Traceback (most recent call last):
  File "/usr/local/lib/python2.7/dist-packages/pip/basecommand.py", line 209, in main
    status = self.run(options, args)
  File "/usr/local/lib/python2.7/dist-packages/pip/commands/install.py", line 317, in run
    prefix=options.prefix_path,
  File "/usr/local/lib/python2.7/dist-packages/pip/req/req_set.py", line 732, in install
    **kwargs
  File "/usr/local/lib/python2.7/dist-packages/pip/req/req_install.py", line 835, in install
    self.move_wheel_files(self.source_dir, root=root, prefix=prefix)
  File "/usr/local/lib/python2.7/dist-packages/pip/req/req_install.py", line 1030, in move_wheel_files
    isolated=self.isolated,
  File "/usr/local/lib/python2.7/dist-packages/pip/wheel.py", line 344, in move_wheel_files
    clobber(source, lib_dir, True)
  File "/usr/local/lib/python2.7/dist-packages/pip/wheel.py", line 304, in clobber
    assert not info_dir, 'Multiple .dist-info directories'
AssertionError: Multiple .dist-info directories
```

This continues to happen for me, on different code bases.

An example from today:

```
$ python3 -m pip install .
Processing /home/bignose/Projects/maildrake/trunk
Collecting docutils (from maildrake===NEXT)
  Using cached docutils-0.13.1-py3-none-any.whl
Collecting fasteners (from maildrake===NEXT)
  Using cached fasteners-0.14.1-py2.py3-none-any.whl
Collecting setuptools (from maildrake===NEXT)
  Using cached setuptools-34.3.2-py2.py3-none-any.whl
Collecting sqlparse (from maildrake===NEXT)
  Using cached sqlparse-0.2.3-py2.py3-none-any.whl
Collecting monotonic>=0.1 (from fasteners->maildrake===NEXT)
  Using cached monotonic-1.3-py2.py3-none-any.whl
Collecting six (from fasteners->maildrake===NEXT)
  Using cached six-1.10.0-py2.py3-none-any.whl
Collecting packaging>=16.8 (from setuptools->maildrake===NEXT)
  Using cached packaging-16.8-py2.py3-none-any.whl
Collecting appdirs>=1.4.0 (from setuptools->maildrake===NEXT)
  Using cached appdirs-1.4.3-py2.py3-none-any.whl
Collecting pyparsing (from packaging>=16.8->setuptools->maildrake===NEXT)
  Using cached pyparsing-2.2.0-py2.py3-none-any.whl
Installing collected packages: docutils, monotonic, six, fasteners, pyparsing, packaging, appdirs, setuptools, sqlparse, maildrake
Exception:
Traceback (most recent call last):
  File "/usr/lib/python3/dist-packages/pip/basecommand.py", line 215, in main
    status = self.run(options, args)
  File "/usr/lib/python3/dist-packages/pip/commands/install.py", line 360, in run
    prefix=options.prefix_path,
  File "/usr/lib/python3/dist-packages/pip/req/req_set.py", line 784, in install
    **kwargs
  File "/usr/lib/python3/dist-packages/pip/req/req_install.py", line 851, in install
    self.move_wheel_files(self.source_dir, root=root, prefix=prefix)
  File "/usr/lib/python3/dist-packages/pip/req/req_install.py", line 1064, in move_wheel_files
    isolated=self.isolated,
  File "/usr/lib/python3/dist-packages/pip/wheel.py", line 345, in move_wheel_files
    clobber(source, lib_dir, True)
  File "/usr/lib/python3/dist-packages/pip/wheel.py", line 305, in clobber
    ', '.join(info_dir))
AssertionError: Multiple .dist-info directories: /home/bignose/.local/lib/python3.5/site-packages/monotonic-1.2.dist-info, /home/bignose/.local/lib/python3.5/site-packages/monotonic-1.3.dist-info

$ ls /home/bignose/.local/lib/python3.5/site-packages/monotonic-*
ls: cannot access '/home/bignose/.local/lib/python3.5/site-packages/monotonic-*': No such file or directory
```

So, for some reason the “Multiple .dist-info directories” message reports two files that *both don't exist* after the command exits. This makes it difficult to know what even went wrong.

This continues to happen even in a clean virtualenv, and when I explicitly clear the Pip cache:

```
(temp_venv) $ rm -r ~/.cache/pip/*

(temp_venv) $ python3 -m pip install ../trunk/dist/maildrake-0.1.2-py2.py3-none-any.whl
Processing /home/bignose/Projects/maildrake/trunk/dist/maildrake-0.1.2-py2.py3-none-any.whl
Requirement already satisfied: semver in ./lib/python3.5/site-packages (from maildrake==0.1.2)
Requirement already satisfied: docutils in ./lib/python3.5/site-packages (from maildrake==0.1.2)
Requirement already satisfied: setuptools in ./lib/python3.5/site-packages (from maildrake==0.1.2)
Collecting fasteners (from maildrake==0.1.2)
  Downloading fasteners-0.14.1-py2.py3-none-any.whl
Requirement already satisfied: sqlparse in ./lib/python3.5/site-packages (from maildrake==0.1.2)
Collecting monotonic>=0.1 (from fasteners->maildrake==0.1.2)
  Downloading monotonic-1.3-py2.py3-none-any.whl
Collecting six (from fasteners->maildrake==0.1.2)
  Downloading six-1.10.0-py2.py3-none-any.whl
Installing collected packages: monotonic, six, fasteners, maildrake
Exception:
Traceback (most recent call last):
  File "/home/bignose/Projects/maildrake/temp_venv/lib/python3.5/site-packages/pip/basecommand.py", line 215, in main
    status = self.run(options, args)
  File "/home/bignose/Projects/maildrake/temp_venv/lib/python3.5/site-packages/pip/commands/install.py", line 360, in run
    prefix=options.prefix_path,
  File "/home/bignose/Projects/maildrake/temp_venv/lib/python3.5/site-packages/pip/req/req_set.py", line 784, in install
    **kwargs
  File "/home/bignose/Projects/maildrake/temp_venv/lib/python3.5/site-packages/pip/req/req_install.py", line 851, in install
    self.move_wheel_files(self.source_dir, root=root, prefix=prefix)
  File "/home/bignose/Projects/maildrake/temp_venv/lib/python3.5/site-packages/pip/req/req_install.py", line 1064, in move_wheel_files
    isolated=self.isolated,
  File "/home/bignose/Projects/maildrake/temp_venv/lib/python3.5/site-packages/pip/wheel.py", line 345, in move_wheel_files
    clobber(source, lib_dir, True)
  File "/home/bignose/Projects/maildrake/temp_venv/lib/python3.5/site-packages/pip/wheel.py", line 305, in clobber
    ', '.join(info_dir))
AssertionError: Multiple .dist-info directories: /home/bignose/Projects/maildrake/temp_venv/lib/python3.5/site-packages/monotonic-1.2.dist-info, /home/bignose/Projects/maildrake/temp_venv/lib/python3.5/site-packages/monotonic-1.3.dist-info
```

There is no information given to help trace where these “multiple .dist-info directories” are coming from, what's making them appear, and what can be done to stop them from appearing.

How can this be diagnosed further?

While #1964 was closed and we now have a better error, I think that this is still a bug.

I have double checked and the wheels have a single .dist-info folder.
I also got the same issue for tar.gz files

This is triggered only when I am using a custom 'build' folder.

The build folder does contain multiple .dist-info, one for the existing version, and one for the version which is going to be installed.

For my project I am pinning everything, so when switching between branches it is common for the CI to reinstall older versions.

An easy way to reproduce it 

```
$ virtualenv pip-build-3564
$ cd pip-build-3564/
$ . bin/activate
$ mkdir build
$ pip install -b build bunch==1.0.0
# All good
$ pip install -b build bunch==1.0.1
AssertionError: Multiple .dist-info directories: /home/adi/tmp/pip-build-3564/lib/python2.7/site-packages/bunch-1.0.0.dist-info, /home/adi/tmp/pip-build-3564/lib/python2.7/site-packages/bunch-1.0.1.dist-info
```

I have standard Ubuntu 16.04-x64
```
$ python --version
Python 2.7.12
$ pip --version
pip 9.0.1
```

If I remove the build folder, all is fine.

Thanks!

Thanks @pfmoore  for the explanation.

But if pip install is not expected to clean the build directory, then why do we have this option:

```
  --no-clean                  Don't clean up build directories.
```

I have never used incremental builds, but my expectation is that for incremental build you will use `--build-directory` together with `--no-clean`

When the `--no-clean` flag is omitted, based on the current documentation I am expecting that the build directory to be cleaned. 

I am triggering this error with a happy path. The initial pip command is successfully executed and I am not passing the `--no-clean` flag... but the next call will fail

```
$ pip install -b build bunch==1.0.0
# All good
$ pip install -b build bunch==1.0.1
AssertionError: Multiple .dist-info directories: /home/adi/tmp/pip-build-3564/lib/python2.7/site-packages/bunch-1.0.0.dist-info, /home/adi/tmp/pip-build-3564/lib/python2.7/site-packages/bunch-1.0.1.dist-info
```

--- a/src/pip/_internal/cmdoptions.py
+++ b/src/pip/_internal/cmdoptions.py
@@ -440,7 +440,8 @@ build_dir = partial(
     help='Directory to unpack packages into and build in. Note that '
          'an initial build still takes place in a temporary directory. '
          'The location of temporary directories can be controlled by setting '
-         'the TMPDIR environment variable (TEMP on Windows) appropriately.'
+         'the TMPDIR environment variable (TEMP on Windows) appropriately. '
+         'When passed, build directories are not cleaned in case of failures.'
 )  # type: Any
 
 ignore_requires_python = partial(

